hat/backends/ptx/src/main/java/hat/backend/PTXCodeBuilder.java

```diff
@@ -0,0 +1,665 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package hat.backend;
+
+import hat.ifacemapper.Schema;
+import hat.optools.*;
+import hat.text.CodeBuilder;
+import hat.util.StreamCounter;
+
+import java.lang.reflect.code.Block;
+import java.lang.reflect.code.Op;
+import java.lang.reflect.code.TypeElement;
+import java.lang.reflect.code.Value;
+import java.lang.reflect.code.op.CoreOp;
+import java.lang.reflect.code.op.ExtendedOp;
+import java.lang.reflect.code.type.JavaType;
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+public class PTXCodeBuilder extends CodeBuilder<PTXCodeBuilder> {
+
+    Map<Value, PTXRegister> varToRegMap;
+    List<String> params;
+    Map<String, Block.Parameter> paramMap;
+    Map<Field, PTXRegister> fieldToRegMap;
+
+    HashMap<PTXRegister.Type, Integer> ordinalMap;
+
+    PTXRegister returnReg;
+    private int addressSize;
+
+    public enum Field {
+        NTID_X ("ntid.x", false),
+        CTAID_X ("ctaid.x", false),
+        TID_X ("tid.x", false),
+        KC_X ("x", false),
+        KC_ADDR("kc", true),
+        KC_MAXX ("maxX", false);
+
+        private final String name;
+        private final boolean destination;
+
+        Field(String name, boolean destination) {
+            this.name = name;
+            this.destination = destination;
+        }
+        public String toString() {
+            return this.name;
+        }
+        public boolean isDestination() {return this.destination;}
+    }
+
+    public PTXCodeBuilder(int addressSize) {
+        varToRegMap = new HashMap<>();
+        params = new ArrayList<>();
+        fieldToRegMap = new HashMap<>();
+        paramMap = new HashMap<>();
+        ordinalMap = new HashMap<>();
+        this.addressSize = addressSize;
+    }
+
+    public PTXCodeBuilder() {
+        this(32);
+    }
+
+    public void ptxHeader(int major, int minor, String target, int addressSize) {
+        append(".version ").append(String.valueOf(major)).dot().append(String.valueOf(minor)).nl();
```

Comment:
I see quite a few uses of append in this file.  

Generally when using your own builder 'dialect' (in this case PTXBuilder), consider each use of raw append and determine whether you should add a specific build method. 

Also don't include '.' and ' ' (space) in strings  (as in append(".version ") prefer explicit dot() and space()

So maybe you add a specific version(), major(int) and minor(int) so your code would be 

dot().version().space().major(major).dot().minor(minor).nl()